### PR TITLE
implement "disconnectTip" in BlockTree

### PR DESCRIPTION
**Pull request by Mr-Leshiy**

### CODE REVIEW

File: include/veriblock/blockchain/blocktree.hpp [hunk 1]
```diff
@@ -262,10 +343,14 @@ struct BlockTree {
                                   index_t& indexNew) {
     if (currentBest.tip() == nullptr ||
         currentBest.tip()->chainWork < indexNew.chainWork) {
+      auto prevTip = currentBest.tip();
       currentBest.setTip(&indexNew);
+      addForkCandidate(prevTip, &indexNew);
```
Comment (blackyblack) [inline include/veriblock/blockchain/blocktree.hpp]: oldCandidate is a second parameter. Why do you put new tip as an oldCandidate?

File: src/blockchain/pop/vbk_block_tree.cpp [hunk 2]
```diff
@@ -139,12 +139,15 @@ void VbkBlockTree::determineBestChain(Chain<block_t>& currentBest,
 
   if (result > 0) {
     // other chain won!
-    return currentBest.setTip(&indexNew);
+    auto prevTip = currentBest.tip();
+    currentBest.setTip(&indexNew);
+    return addForkCandidate(prevTip, &indexNew);
```
Comment (blackyblack) [inline src/blockchain/pop/vbk_block_tree.cpp]: Also prevTip as a new candidate.

File: include/veriblock/blockchain/blocktree.hpp [hunk 3]
```diff
@@ -109,10 +109,54 @@ struct BlockTree {
...
+    if (blockIndex == nullptr) {
+      return state.Invalid("invalidateBlockByHash()",
+                           "bad-block-hash",
+                           "can not find block by this hash");
+    }
+
+    for (auto it = fork_candidates_.begin(); it != fork_candidates_.end();) {
+      index_t* forkIndex = *it;
+      index_t* prevIndex = nullptr;
+      if (forkIndex->getAncestor(blockIndex->height) == blockIndex) {
```
[the remark this answers is not in the document]
Comment (blackyblack) [reply include/veriblock/blockchain/blocktree.hpp]: I think there is another issue as well. We might erase a branch having a common block with another branch. This way another branch will become orphaned.

[the remark this answers is not in the document]
Comment (blackyblack) [reply include/veriblock/blockchain/blocktree.hpp]: I would recommend to mark branches containing block that's being deleted first. And physically erase branches on the next step.

File: include/veriblock/blockchain/blocktree.hpp [hunk 4]
```diff
@@ -109,10 +109,32 @@ struct BlockTree {
     return acceptBlock(block, state, true);
   }
 
+  void invalidateBlockByHash(const hash_t& blockHash) {
+    index_t* blockIndex = getBlockIndex(blockHash);
```
Comment (Warchant) [inline include/veriblock/blockchain/blocktree.hpp]: here you should check that block index is not nullptr.
`getBlockIndex` returns nullptr when it can't find a block with this hash

File: include/veriblock/blockchain/blocktree.hpp [hunk 5]
```diff
@@ -148,15 +169,74 @@ struct BlockTree {
       current->height = current->pprev->height + 1;
       current->chainWork = current->pprev->chainWork + getBlockProof(block);
     } else {
-      current->height = 0;
```
Comment (Warchant) [inline include/veriblock/blockchain/blocktree.hpp]: why did you remove that?

File: include/veriblock/blockchain/blocktree.hpp [hunk 6]
```diff
@@ -148,15 +169,74 @@ struct BlockTree {
...
     return current;
   }
 
+  void invalidateBlockFromChain(Chain<Block>& chain, index_t** block) {
+    if (*block == nullptr) {
+      return;
+    }
+
+    if (!chain.contains(*block) &&
+        chain.tip()->getAncestor((*block)->height) != *block) {
```
Comment (Warchant) [inline include/veriblock/blockchain/blocktree.hpp]: why do you need this check here?
isn't `contains` enough?

File: include/veriblock/blockchain/chain.hpp [hunk 7]
```diff
@@ -57,9 +57,11 @@ struct Chain {
     return (*this)[index->height + 1];
   }
 
-  height_t size() const { return (height_t)chain.size() + startHeight_; }
+  height_t height() const { return (height_t)chain.size() + startHeight_ - 1; }
```
Comment (Warchant) [inline include/veriblock/blockchain/chain.hpp]: when chain is empty, and startHeight is 0, height will point to -1, which is wrong. Better use `size() -1` in places where you need `height` of the chain

File: include/veriblock/blockchain/miner.hpp [hunk 8]
```diff
@@ -15,7 +14,8 @@ struct Miner {
   using merkle_t = decltype(Block::merkleRoot);
   using index_t = BlockIndex<Block>;
 
-  Miner(std::shared_ptr<ChainParams> params) : params_(std::move(params)) {}
+  Miner(std::shared_ptr<ChainParams> params, uint32_t startTime)
```
Comment (Warchant) [inline include/veriblock/blockchain/miner.hpp]:
```suggestion
  Miner(std::shared_ptr<ChainParams> params, uint32_t startTime = currentTimestamp4())
```

File: include/veriblock/blockchain/blocktree.hpp [hunk 9]
```diff
@@ -109,10 +109,32 @@ struct BlockTree {
     return acceptBlock(block, state, true);
   }
 
+  void invalidateBlockByHash(const hash_t& blockHash) {
+    index_t* blockIndex = getBlockIndex(blockHash);
+
+    for (auto chain_it = fork_chains_.begin();
+         chain_it != fork_chains_.end();) {
+      invalidateBlockFromChain(*chain_it, &blockIndex);
```
Comment (blackyblack) [inline include/veriblock/blockchain/blocktree.hpp]: In case common block of two branches is deleted one of the branches becomes orphaned.

File: include/veriblock/blockchain/blocktree.hpp [hunk 10]
```diff
@@ -148,15 +169,74 @@ struct BlockTree {
...
+      if (chain_it->tip() == newCandidate) {
+        isAdded = true;
+      }
+
+      if (chain_it->tip() == oldCandidate) {
+        chain_it = fork_chains_.erase(chain_it);
+        continue;
+      }
+      ++chain_it;
+    }
```
Comment (blackyblack) [inline include/veriblock/blockchain/blocktree.hpp]: It seems there are two not connected actions: deleting oldCandidate and adding newCandidate. I think they should reside in two different functions.

File: include/veriblock/blockchain/blocktree.hpp (hunk 6 above)
Comment (Mr-Leshiy) [reply include/veriblock/blockchain/blocktree.hpp]: no it is not enough, because fork chains dont have all blocks, and it is on that case that we want do remove the old block from the main chain, and as a result we have to remove the parts of the forks.

File: include/veriblock/blockchain/chain.hpp (hunk 7 above)
Comment (Mr-Leshiy) [reply include/veriblock/blockchain/chain.hpp]: I am not sure, because the meaning of the size is not the similar with the realisation, it is not return the size of the chain.

File: include/veriblock/blockchain/blocktree.hpp (hunk 6 above)
Comment (Warchant) [reply include/veriblock/blockchain/blocktree.hpp]: instead of getAncestor use `activeChain.contains(*block)`

File: include/veriblock/blockchain/blocktree.hpp [hunk 11]
```diff
@@ -147,10 +177,93 @@ struct BlockTree {
...
+      return;
+    }
+
+    // find block from the main chain
+    index_t* workBlock = newCandidate;
+    if (forkChainStart == fork_chains_.end()) {
+      for (; !activeChain_.contains(workBlock->pprev);
+           workBlock = workBlock->pprev)
+        ;
+    }
```
Comment (Warchant) [inline include/veriblock/blockchain/blocktree.hpp]: use `auto* forkBlock = chain.findFork(tip)`

File: include/veriblock/blockchain/blocktree.hpp [hunk 12]
```diff
@@ -107,10 +107,58 @@ struct BlockTree {
...
+    // and if block F is being removed, and chain 'F - G - H - I' will be
+    // processed firstly, we will remove block 'G' which is needed for the
+    // another chain, so that chain must be processed firstly
+
+    std::sort(
+        fork_chains_.begin(),
+        fork_chains_.end(),
+        [](const Chain<Block>& chain1, const Chain<Block>& chain2) -> bool {
+          return chain1.getStartHeight() > chain2.getStartHeight();
+        });
```
Comment (Warchant) [inline include/veriblock/blockchain/blocktree.hpp]: use std::map to store forks. it will maintain sorted order (by height)
